### DISCUSSION TITLE
One-to-many relationship but only selecting the first row (by order)

### DISCUSSION THREAD

**Discussion by ported-pw** [sqlalchemy/sqlalchemy#9021] [Usage Questions] [answered]
Hi!

I hope my explanation of my goal will be clear enough.
Basically, I am currently trying to convert an old application to SQLAlchemy 1.4 (with AsyncIO, so with the 2.0 query style, and then onto 2.0 beta afterwards) that previously used a lot of dynamic loading which now isn't possible the same way and/or I want to make things faster anyway.

I used to have models like this:
```python
class UserName(Base):
   id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
   user__id = Column(UUID(as_uuid=True), ForeignKey("user.id"), nullable=False)
   
   user = relationship("User", back_populates="names")

class User(Base):
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    
    names = relationship("UserName", order_by="UserName.first_seen.desc()")
    
    @property
    def latest_name(self):
        return self.names.first()
```
The `latest_name` property would be used to fetch only the latest name in the related collection (by an API schema like Pydantic or previously Marshmallow).
Obviously, this is 1) very ineffecient for large result sets and 2) does not work anymore.

I assume the SQL way to accomplish a joined load of `latest_name` would be to have a correlated subquery with `LIMIT 1` and select from that.
Is there any way to transparently configure a model to achieve something like a join-loaded `latest_name` (via subquery or otherwise) in this situation?
If not, what would be the next-prettiest way to do this given that I want to load it into a Pydantic model with `Model.from_orm()`?

Thanks!

**Comment by zzzeek** [comment]
yes the LIMIT 1 is not the most portable way to do this, the best pattern we have is using window functions (which are universally available now), see https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#row-limited-relationships-with-window-functions

**Reply by ported-pw** [reply]
Thank you so much, that's exactly what I need and I got it working with `index == 1` and `uselist=False` on the relationship!
I am a bit annoyed that I googled my eyes out for two hours and it was right there, perfectly documented as always.
Thanks for your work on SQLAlchemy and the fast help :)

**Reply by ported-pw** [reply]
Hm, it works well enough when querying single `User`s (as in my example) but falls apart pretty much instantly when I try selecting by anything but `User.id`. PostgreSQL then, instead of using an `Index Scan Backward` for the `WindowAgg` part of the query, uses a `Incremental Sort` that even using an index scan is insanely slow on my ~35M row table.
Do you maybe know of a better aggregation function to use? (I know it's not an SQLAlchemy question anymore at this point :D)

**Reply by zzzeek** [reply]
this is relationship loading that loads whole objects, so i dont understand what you mean by "selecting anything but User.id".    Can you provide a complete example?

**Reply by ported-pw** [reply]
Yes, I will try. This would be my initial example, plus:
- Adding table names and indexes
- Adding the `UserName.first_seen` column that was missing in the example
- Using the suggested window function instead of the dynamic load
- Adding another model / one-to-many relationship to `User` that is joined in and filtered by

The relevant part of the real application looks quite close to this (but I could have made mistakes distilling it down):
```python
class Infraction(Base):
    __tablename__ = "infraction"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user__id = Column(UUID(as_uuid=True), ForeignKey("user.id"), nullable=False)
    
    created = Column(TIMESTAMP, index=True, default=datetime.utcnow, nullable=False)
    cleared = Column(BOOLEAN, default=False, nullable=False)
    
    user = relationship("User", back_populates="infractions")
    
    __table_args__ = (
        # combined index over user id, cleared flag, and sorted created date
        Index("infraction_user__id_cleared_created", user__id, cleared, created.desc()),
    )


class UserName(Base):
   __tablename__ = "user_name"
    
   id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
   user__id = Column(UUID(as_uuid=True), ForeignKey("user.id"), nullable=False)
   first_seen = Column(TIMESTAMP, nullable=False, default=datetime.utcnow)
   
   user = relationship("User", back_populates="names")
   
   __table_args__ = (
        # combined index over user id and first seen
        Index("user_name_user__id_first_seen_idx", user__id, first_seen.desc()),
   )
   

class User(Base):
    __tablename__ = "user"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    
    infractions = relationship("Infraction", order_by="Infraction.created.desc()")
    names = relationship("UserName", order_by="UserName.first_seen.desc()")

    
latest_name_partition = select(
    UserName,
    func.row_number().over(order_by=UserName.last_seen.desc(), partition_by=UserName.user__id).label("index"),
).alias()

latest_name_partitioned = aliased(UserName, latest_name_partition)

User.latest_name = relationship(
    latest_name_partitioned,
    primaryjoin=and_(latest_name_partitioned.user__id == User.id, latest_name_partition.c.index == 1),
    uselist=False,
    lazy="joined",
)
```

Now, if I do the following query against this, filtering by `User.id` and ignoring `infractions`:
```python
select(User).filter(User.id == "8c5c6bb5-ecae-4289-b8b8-ab80d45c604c")
```

This will create the following query:
```sql
SELECT user.id,
       anon_1.id AS id_1,
       anon_1.user__id,
       anon_1.first_seen,
FROM user
         LEFT OUTER JOIN
     (SELECT user_name.id                           AS id,
             user_name.user__id                     AS user__id,
             user_name.first_seen                   AS first_seen,
             row_number() OVER (PARTITION BY user_name.user__id
                 ORDER BY user__id.first_seen DESC) AS INDEX
      FROM user_name) AS anon_1 ON anon_1.user__id = user.id
         AND anon_1.index = 1
WHERE user.id = '8c5c6bb5-ecae-4289-b8b8-ab80d45c604c'
```

Which results in this query plan on PostgreSQL:
```
Nested Loop Left Join  (cost=1.00..284.21 rows=1 width=165)
  Join Filter: (anon_1.user__id = user.id)
  ->  Index Scan using user_pkey on user  (cost=0.43..8.45 rows=1 width=85)
        Index Cond: (id = '8c5c6bb5-ecae-4289-b8b8-ab80d45c604c'::uuid)
  ->  Subquery Scan on anon_1  (cost=0.56..275.74 rows=1 width=80)
        Filter: (anon_1.index = 1)
        ->  WindowAgg  (cost=0.56..274.91 rows=67 width=88)
              ->  Index Scan Backward using user_name_user__id_first_seen_idx on user_name  (cost=0.56..273.73 rows=67 width=80)
                    Index Cond: (user__id = '8c5c6bb5-ecae-4289-b8b8-ab80d45c604c'::uuid)
```
which as you can see is very efficient/fast and exactly what I need.

Now we complicate things a bit, and instead of filtering `User` by `User.id`, we join in `infractions` and filter by that:
```python
(
    select(User)
    .join(User.infractions)
    .filter(Infraction.cleared.is_(False))
    .order_by(Infraction.created.desc())
)
```

This will create the following query:
```sql
SELECT user.id,
       anon_1.id AS id_1,
       anon_1.user__id,
       anon_1.first_seen,
FROM user
         JOIN infraction ON user.id = infraction.user__id
         LEFT OUTER JOIN (SELECT user_name.id                                                               AS id,
                                 user_name.user__id                                                         AS user__id,
                                 user_name.first_seen                                                       AS first_seen,
                                 row_number()
                                 OVER (PARTITION BY user_name.user__id ORDER BY user_name.last_seen DESC) AS index
                          FROM user_name) AS anon_1 ON anon_1.user__id = user.id AND anon_1.index = 1
WHERE infraction.cleared IS false
ORDER BY infraction.created DESC
```

Which will lead to the now significantly less favourable query plan:
```
Nested Loop Left Join  (cost=89.44..45939249.13 rows=73 width=173)
  Join Filter: (anon_1.user__id = user.id)
  ->  Nested Loop  (cost=0.72..3068.22 rows=73 width=93)
        ->  Index Scan Backward using ix_infraction_created on infraction  (cost=0.29..2451.37 rows=73 width=24)
              Filter: ((cleared IS FALSE) AND (ban__id IS NULL))
        ->  Index Scan using user_pkey on user  (cost=0.43..8.45 rows=1 width=85)
              Index Cond: (id = infraction.user__id)
  ->  Materialize  (cost=88.71..45751611.78 rows=168942 width=80)
        ->  Subquery Scan on anon_1  (cost=88.71..45750767.07 rows=168942 width=80)
              Filter: (anon_1.index = 1)
              ->  WindowAgg  (cost=88.71..45328413.17 rows=33788312 width=88)
                    ->  Incremental Sort  (cost=88.71..44737117.71 rows=33788312 width=80)
                          Sort Key: user_name.user__id, user_name.last_seen DESC
                          Presorted Key: user_name.user__id
                          ->  Index Scan using user_name_user__id_last_seen_idx on user_name  (cost=0.56..42575449.11 rows=33788312 width=80)
```

As I understand it, this basically sorts the entire `user_name` table first at immense cost (so they query probably won't complete any time soon) because there's nothing in the `WHERE` clause anymore that is directly used for the `JOIN ON` (or rather, nothing is used in the `WHERE` clause that is also present on `user_name`).

I suppose what it would have to do is run the query with the joined `infraction` table first, and then a subquery for `user_name` each second
(which would be back to a correlated subquery instead of the window function, which I then wouldn't know how to configure onto a relationship).

Again, thanks for your help! :)

**Reply by zzzeek** [reply]
if there is always at least one UserName per User, you could turn that LEFT OUTER JOIN into a JOIN by using joinedload(innerjoin=True).

Or, I would just eliminate the issue by using the selectinload() loader intead of joinedload() so you aren't stuck with that LEFT OUTER JOIN, a separate query at most one inner join just the straight, very fast query will be emitted.

There's other ways to get "the latest X for each Y", but they would all do similar things.

For example, heres the example from https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#row-limited-relationships-with-window-functions using plain group_by instead:

```py
partition = select(
    B, func.min(B.id)
).group_by(B.a_id).alias()

b_ids_we_want = select(func.min(B.id).label("id")).group_by(B.a_id).alias()

bs_we_want = select(B).join(b_ids_we_want, b_ids_we_want.c.id == B.id).subquery()
partitioned_b = aliased(B, bs_we_want)

A.partitioned_bs = relationship(
    partitioned_b
)

```

a joinedload() query looks like this then:

```py
SELECT a.id AS a_id, anon_1.id AS anon_1_id, anon_1.a_id AS anon_1_a_id 
FROM a LEFT OUTER JOIN (SELECT b.id AS id, b.a_id AS a_id 
FROM b JOIN (SELECT min(b.id) AS id 
FROM b GROUP BY b.a_id) AS anon_2 ON anon_2.id = b.id) AS anon_1 ON a.id = anon_1.a_id
```

basically the same thing.  maybe postgresql would optimize it differently.

**Reply by CaselIT** [reply]
Probably this would require a `JOIN LATERAL` to provide a better query plan, since you could filter for the user id inside the selected join (avoiding also the partition by in the over function)
I'm not sure if joinedload support it

**Reply by zzzeek** [reply]
selectinload is the way to go here

**Reply by CaselIT** [reply]
I'm not sure if it helps, since the where is still on the outside of the subquery. Pg should figure it out and apply it on the inside (if partition by is used) but that may not happen

**Reply by ported-pw** [reply]
Hey and thanks for your response, hope you had a good holiday time!

> if there is always at least one UserName per User, you could turn that LEFT OUTER JOIN into a JOIN by using joinedload(innerjoin=True).

Unfortunately that's not the case, the naming in the example is a bit lacking here suggesting otherwise.

> Or, I would just eliminate the issue by using the selectinload() loader intead of joinedload() so you aren't stuck with that LEFT OUTER JOIN, a separate query at most one inner join just the straight, very fast query will be emitted.

True! I entirely forgot about trying to apply different loader strategies to what is now a proper relationship.
I've tried it, and it's a lot better, but because the query still uses a `JOIN` on `User` instead of selecting directly from `UserName` instead, it ends up performing a sequential scan on `UserName` to sort it for the aggregation anyway.

Here's the abridged query generated by the load part of the query using `selectinload` on the relationship in my example:
```python
User.latest_name = relationship(
    latest_name_partitioned,
    primaryjoin=and_(latest_name_partitioned.user__id == User.id, latest_name_partition.c.index == 1),
    uselist=False,
    lazy="selectin",
)
```

```sql
EXPLAIN SELECT user_1.id           AS user_1_id,
       anon_1.id             AS anon_1_id,
       anon_1.user__id       AS anon_1_user__id,
       anon_1.first_seen     AS anon_1_first_seen
FROM player AS user_1
         JOIN (SELECT user_name.id                                                                            AS id,
                      user_name.user__id                                                                      AS user__id,
                      user_name.first_seen                                                                    AS first_seen,
                      row_number() OVER (PARTITION BY user_name.user__id ORDER BY user_name.last_seen DESC) AS index
               FROM user_name) AS anon_1 ON anon_1.user__id = user_1.id AND anon_1.index = 1
WHERE user_1.id IN ('77abd848-254a-4f10-a428-e01623c03778', '2e1a575c-01aa-4010-b3e4-bac0b94bb517')
```
Which results in this query plan:
```
Merge Join  (cost=6716444.11..7846732.45 rows=1 width=96)
  Merge Cond: (anon_1.user__id = user_1.id)
  ->  Subquery Scan on anon_1  (cost=6716443.67..7846284.98 rows=173822 width=80)
        Filter: (anon_1.index = 1)
        ->  WindowAgg  (cost=6716443.67..7411730.63 rows=34764348 width=88)
              ->  Sort  (cost=6716443.67..6803354.54 rows=34764348 width=80)
                    Sort Key: user_name.user__id, user_name.first_seen DESC
                    ->  Seq Scan on user_name  (cost=0.00..817309.48 rows=34764348 width=80)
  ->  Index Only Scan using player_pkey on player user_1  (cost=0.43..12.90 rows=2 width=16)
        Index Cond: (id = ANY ('{77abd848-254a-4f10-a428-e01623c03778,2e1a575c-01aa-4010-b3e4-bac0b94bb517}'::uuid[]))
```

I posted what I think is a decent solution separately: https://github.com/sqlalchemy/sqlalchemy/discussions/9021#discussioncomment-4504099

**Comment by ported-pw** [comment, marked as answer]
### A decent solution for this exact case (I think, I could be doing something wrong)

Following up on @zzzeek comments here.

After searching a bit, the best raw SQL solution seem to be just using a plain `SELECT DISTINCT ON` with an `ORDER BY`.

So I've configured my relationship like this (see the example models in my comment: https://github.com/sqlalchemy/sqlalchemy/discussions/9021#discussioncomment-4488523):
```python
latest_name_subq = (
    select(UserName).distinct(UserName.user__id).order_by(UserName.user__id, UserName.first_seen.desc()).alias()
)
latest_name_alias = aliased(UserName, latest_name_subq)
User.latest_name = relationship(
    latest_name_alias,
    uselist=False,
    lazy="selectin",
)
```
with while `joined` creates some horribly query plan, **works beautifully** with `selectin` loading using just a few index scans and my HTTP request returning in <200ms.

Thanks again for your help @zzzeek !

**Comment by Joaquim-Ferrer** [comment]
@ported-pw when I try your solution, sqlalchemy seems to load all classes at once and it doesn't work because python stops recognising Class names between `"`

Like I have a class:
    
```
class PullRequests
    user: Mapped[Optional["User"]] = relationship(
        "User",
        back_populates="pull_requests",
        primaryjoin="User.external_vcs_id == PullRequest.author_external_id",
        foreign_keys="[PullRequest.author_external_id]",
    ) 
```

And it starts throwing:
"When initializing mapper Mapper[PullRequest(vcs_request)], expression 'User' failed to locate a name ('User'). If this is a class name, consider adding this relationship() to the <class 'enterprise_core.main.dal.db.postgres.orms.pull_request.PullRequest'> class after both dependent classes have been defined."

This only happens after I call the aliased method.

Any ideas?
